### PR TITLE
implement Entry Proxy in Go

### PR DESCRIPTION
**Pull request by starius**
This work is not finished!

Testing:
- **DNS rules for the domain**. I added domain `test-ssl.pasta.cf` with A record "127.0.0.1" and TXT record "onion=pokodugigmf66twh.onion".
- **Server**. Onion site pokodugigmf66twh.onion was created (`HiddenServicePort 4218 127.0.0.1:4218`). For testing purposes I used port 4218 rather than 443. `nginx` listens on port 4218 and provides access to back-end of my pastebin like site pasta.cf.
- **Local setup**. Compiled tool `entry_proxy` with `go get && go install`. Run: `entry_proxy -listen-on 127.0.0.1:4218 -onion-port 4218`. In curl `curl --insecure https://test-ssl.pasta.cf:4218`. It works!

My next step is to run `entry_proxy` on some public VPS and to set A record of the domain to IP address of that VPS. By the way, we can add IPv6 addresses as well.

Previous repository for this code: https://github.com/starius/epog

### CODE REVIEW

File: entry_proxy/main.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        if err != nil {
+            log.Printf("Finished writting: %s", err)
+            break
+        }
+    }
+}
+
+func processRequest(clientConn net.Conn) {
+    defer clientConn.Close()
+    hostname, clientConn, err := sni.ServerNameFromConn(clientConn)
```
Comment (david415) [inline entry_proxy/main.go:64]: yes! all parsing done here

File: entry_proxy/main.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    connection, err := dialer.Dial("tcp", targetServer)
+    return connection, err
+}
+
+func netCopy(from, to net.Conn, finished chan<- struct{}) {
+    defer func() {
+        finished<-struct{}{}
+    }()
+    buffer := make([]byte, *bufferSize)
+    for {
```
Comment (david415) [inline entry_proxy/main.go:48]: at least for unit test purposes i think it's better to have a shutdown code path. currently it only breaks if a read or write error occur.

File: entry_proxy/main.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        from.SetReadDeadline(time.Now().Add(*timeout))
```
Comment (david415) [inline entry_proxy/main.go]: setting the deadline can cause this code to take lots of cpu time. does it work without setting the deadline? why is this needed?

Comment (starius) [inline entry_proxy/main.go]: It works without setting the deadline. The timeout was introduced to 
deal with peers who do not close the connection properly. They can 
exhaust all sockets and stop normal operation of the service. 
However this code was removed, because we have to confirm this 
problem before fixing it.

File: entry_proxy/main.go (hunk 2 above)
Comment (starius) [inline entry_proxy/main.go:48]: EOF is also an error for `Read` and `Write` methods. I think, it can be produced in a test.

File: entry_proxy/main.go (hunk 1 above)
Comment (starius) [inline entry_proxy/main.go:64]: By the way, tests of https://github.com/polvi/sni fail with the following error:
https://gist.github.com/108ba815eb3ecb3fdcdaa245194a8b46

Comment (starius) [inline entry_proxy/main.go:64]: Fixed in https://github.com/polvi/sni/pull/1

**Comment by starius** [comment]
I started `entry_proxy` on 2002:5e66:35b2::1 and 94.102.53.178 on port 443.
Domain www.pasta.cf was configured to have A=94.102.53.178 and AAAA=2002:5e66:35b2::1.
(Non-www domain configured for direct access, because this domain is not really secret.)

To deal with ban by the DNS server for frequent DNS requests and to decrease latency, I installed caching DNS server `unbound` and put `nameserver 127.0.0.1` to `/etc/resolv.conf`. Go seems to re-read this file every 5 seconds (see `dnsclient_unix.go`).

Tor was built in Tor2Web mode according to [Tor2Web guide](https://github.com/globaleaks/Tor2web/wiki/Installation-Guide#build-tor-with-tor2web-mode-and-some-patches).

It seems to work.
